File: Amazon/Onsite/C++/Design Log Storage System.cpp
class LogSystem {
public:
    LogSystem() {
        
    }
    
    void put(int id, string timestamp) {
        logs.emplace_back(id, timestamp);
    }
    
    vector<int> retrieve(string s, string e, string gra) {
        vector<int> res;
        int idx = indexes[gra];
        for(auto log : logs)
        {    
            string timestamp = log.second.substr(0, idx);
            if(s.substr(0, idx) <= timestamp &&  timestamp <= e.substr(0, idx))
                res.push_back(log.first);
        }
        
        return res;
    }
    
private:
    vector<pair<int, string>> logs;
    unordered_map<string, int> indexes {{"Year", 5} , {"Month", 8}, {"Day", 11}, 
                                        {"Hour", 14}, {"Minute", 17}, {"Second", 20}};
};

File: Amazon/Onsite/Python/Design Log Storage System.py
class LogSystem:

    def __init__(self):
        # list is ok
        self.logs = []
        self.g = {'Year': 0, 'Month':1, 'Day':2, 'Hour': 3, 'Minute':4, 'Second':5}
    def put(self, id: int, timestamp: str) -> None:
        self.logs.append([id, timestamp])

    def retrieve(self, s: str, e: str, gra: str) -> List[int]:
        res = []
        s = s.split(':')
        s = s[:self.g[gra]+1]
        e = e.split(':')
        e = e[:self.g[gra]+1]
        for log in self.logs:
            if self.check(log[1], s, e, gra):
                res.append(log[0])
        return res
    
    def check(self, time, s, e, gra):
        time = time.split(':')
        time = time[:self.g[gra]+1]
        return s <= time <= e